src/devices/conv.rs

```diff
@@ -48,28 +55,46 @@ impl<const S: usize, const P: usize> DeviceConv2D<S, P> for Cpu {
         bias: &[f32; O],
         out: &mut [[[f32; (W + 2 * P - K) / S + 1]; (H + 2 * P - K) / S + 1]; O],
     ) {
+        let mut patches: Box<
+            [[[[[f32; (W + 2 * P - K) / S + 1]; (H + 2 * P - K) / S + 1]; K]; K]; C],
+        > = Self::zeros();
+
         for c in 0..C {
-            for oc in 0..O {
-                for oh in 0..((H + 2 * P - K) / S + 1) {
-                    for ow in 0..((W + 2 * P - K) / S + 1) {
-                        let o = &mut out[oc][oh][ow];
-                        let mut tmp = 0.0;
-                        for k1 in 0..K {
-                            let y = (oh * S + k1).checked_sub(P);
-                            for k2 in 0..K {
-                                let x = (ow * S + k2).checked_sub(P);
-                                if let Some((y, x)) = y.zip(x) {
-                                    if y < H && x < W {
-                                        tmp += weight[oc][c][k1][k2] * img[c][y][x];
-                                    }
-                                }
+            for k1 in 0..K {
+                for k2 in 0..K {
+                    for oh in 0..(H + 2 * P - K) / S + 1 {
+                        for ow in 0..(W + 2 * P - K) / S + 1 {
+                            let y = (oh * S + k1).wrapping_sub(P);
+                            let x = (ow * S + k2).wrapping_sub(P);
+                            if y < H && x < W {
+                                patches[c][k1][k2][oh][ow] = img[c][y][x];
                             }
                         }
-                        *o += tmp;
                     }
                 }
             }
         }
+
+        // (O, C * K * K) * (C * K * K, OH * OW) = (O, OH * OW)
+        let m = O;
+        let k = C * K * K;
+        let n = ((H + 2 * P - K) / S + 1) * ((W + 2 * P - K) / S + 1);
+        let a = weight.as_ptr() as *const f32;
+        let b = patches.as_ptr() as *const f32;
+        let c = out.as_mut_ptr() as *mut f32;
+        #[cfg(not(feature = "cblas"))]
+        unsafe {
+            matrixmultiply::sgemm(
+                m, k, n, 1.0, a, k as isize, 1, b, n as isize, 1, 1.0, c, n as isize, 1,
+            )
+        }
+
+        #[cfg(feature = "cblas")]
+        unsafe {
+            let (m, n, k) = (m as libc::c_int, n as libc::c_int, k as libc::c_int);
+            sgemm(RowMajor, NoTr, NoTr, m, n, k, 1.0, a, k, b, n, 1.0, c, n)
+        }
```

Comment:
Noting that this is copied from the matmul implementation - using Cpu::mm requires casting the inputs to 2d arrays, which in turn requires generic_const_expr bounds added to the trait, which I wanted to avoid.